fix(macos): Build deps from source to allow backwards compatibility

anthonybilinski opened the pull request:
- [x] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/6438)
<!-- Reviewable:end -->

nurupo commented on buildscripts/build_vpx_windows.sh:
```diff
@@ -17,45 +17,49 @@
+    TARGET="--target=x86-win32-gcc"
+    CROSS_ARG="${ARCH}-w64-mingw32-"
+elif [ "${ARCH}" == "macos" ]; then \
+    ARCH_FLAGS="-fno-asynchronous-unwind-tables"
```
Is it really needed for macOS or did you just blindly copy it from the Windows build script?

nurupo replied:
Just to clarify, during the CI refactor (PR https://github.com/qTox/qTox/pull/6413) a comment was dropped explaining why this compiler flag was needed for the Windows build:

```sh
# There is a bug in gcc that breaks avx512 on 64-bit Windows https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54412
# VPX fails to build due to it.
# This is a workaround as suggested in https://stackoverflow.com/questions/43152633
VPX_CFLAGS="-fno-asynchronous-unwind-tables"
```

(This comment is getting restored in PR https://github.com/qTox/qTox/pull/6458)

What the comment says is that that cflag is in there just to avoid a Windows/MinGW specific bug in gcc.

Having the workaround for a Windows/MinGW specific bug applied to macOS looks rather suspicious, thus my earlier question.

anthonybilinski replied:
Yeah reasonable, realized after reviewing the PR replacing the comment. Will test and likely remove, I did just carry it over from the Windows build here.

macOS build breaks on two of the deps due to some rpath difference, for some reason, so some work left there too, but will come back to this.

anthonybilinski committed:
chore(CI): Rename Windows vpx.patch to vpx-windows.patch

Avoid conflict with different VPX patch needed for macos

anthonybilinski committed:
chore(CI): Add script for handling platform compile args

Deduplicates between all the small independent scripts, and further
deduplicates with incoming macOS configurations.

anthonybilinski committed:
chore(CI): Rename Windows build scripts that will be used commonly by macOS

anthonybilinski committed:
chore(CI): Add macOS VPX patch to allow setting mmacosx-version-min

VPX overrides the user setting by default, preventing a newer macOS from
building for an older macOS.

anthonybilinski committed:
chore(CI): Use platform detection script in Windows dep build scripts

Deduplicates a lot of arg checking, and further deduplicates future macOS args.

anthonybilinski committed:
chore(CI): Move QrEncode build from autotools to cmake

Autotools version failed to configure on macOS, and QrEncode's README says  "If
the configure script does not work well, try to use CMake." Cmake works for
both macOS and Windows.

anthonybilinski committed:
chore(CI): Add support for building cmake deps for macOS

DCMAKE_OSX_DEPLOYMENT_TARGET has no effect when not building for macOS. When
building for macOS, ensures compatibility with macOS versions matching the
specified version by preventing new optimizations or API usages. By default,
compiled software will only be supported on the same version of macOS or newer.

anthonybilinski committed:
chore(CI): Add support for building VPX for macOS

Manually set library id name because it defaults to "libvpx.7.dylib",
for some reason, which is not found by qTox or tests.

anthonybilinski committed:
chore(CI): Add support for building sqlcipher for macOS

anthonybilinski committed:
chore(CI): Add support for building sodium for macOS

anthonybilinski committed:
chore(CI): Add support for building opus for macOS

anthonybilinski committed:
chore(CI): Add support for building openssl for macOS

anthonybilinski committed:
chore(CI): Add support for building openal for macOS

MACOSX_RPATH must be set explicitly to be found. It is default using newer cmake, but OpenAL is using cmake 2.6.

anthonybilinski committed:
chore(CI): Add support for building libexif for macOS

anthonybilinski committed:
chore(CI): Add support for building ffmpeg for macOS

anthonybilinski committed:
chore(CI): Add build script for building all macOS deps

anthonybilinski committed:
chore(CI): Split brewfile between building all deps or only qTox

anthonybilinski committed:
chore(CI): macOS CI script support for both full dep build or just qTox build

(cherry picked from commit bffa83b14fb06c78454bfe42289379de82344512)

anthonybilinski committed:
feat(macOS): Build deps from source, giving backwards compatibility

Keep second CI job building with deps from brew, to test the workflow of a user
building qTox from source, with deps from brew.

Fix #6260

(cherry picked from commit 785ae6874796c2f9a7bfd9b8aa88a1fc34ead3d2)

anthonybilinski commented:
I've split my monster change into (literally) 20 more reasonable commits and rebased.

I'm a big stumped on the macOS version built from source not finding a couple of the dependencies, but finding the rest. It finds all of them at configure time, but then the rpath seems to be incorrectly set in the binary.

anthonybilinski commented:
Follow up to that, my otool output, equivalent to ldd:

```
abilinski@abilinskis-iPro _build % otool -L qtox.app/Contents/MacOS/qtox
qtox.app/Contents/MacOS/qtox:
        /usr/local/opt/qt@5/lib/QtNetwork.framework/Versions/5/QtNetwork (compatibility version 5.15.0, current version 5.15.2)
        /usr/local/opt/qt@5/lib/QtOpenGL.framework/Versions/5/QtOpenGL (compatibility version 5.15.0, current version 5.15.2)
        /usr/local/opt/qt@5/lib/QtSvg.framework/Versions/5/QtSvg (compatibility version 5.15.0, current version 5.15.2)
        /usr/local/opt/qt@5/lib/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.15.0, current version 5.15.2)
        /usr/local/opt/qt@5/lib/QtXml.framework/Versions/5/QtXml (compatibility version 5.15.0, current version 5.15.2)
        /Users/abilinski/builds/qTox/local-deps/lib/libavcodec.58.dylib (compatibility version 58.0.0, current version 58.134.100)
        /Users/abilinski/builds/qTox/local-deps/lib/libavdevice.58.dylib (compatibility version 58.0.0, current version 58.13.100)
        /Users/abilinski/builds/qTox/local-deps/lib/libavformat.58.dylib (compatibility version 58.0.0, current version 58.76.100)
        /Users/abilinski/builds/qTox/local-deps/lib/libavutil.56.dylib (compatibility version 56.0.0, current version 56.70.100)
        /Users/abilinski/builds/qTox/local-deps/lib/libexif.12.dylib (compatibility version 16.0.0, current version 16.4.0)
        @rpath/libqrencode.4.dylib (compatibility version 4.0.0, current version 4.1.1)
        /Users/abilinski/builds/qTox/local-deps/lib/libsodium.23.dylib (compatibility version 27.0.0, current version 27.0.0)
        /Users/abilinski/builds/qTox/local-deps/lib/libswscale.5.dylib (compatibility version 5.0.0, current version 5.9.100)
        /Users/abilinski/builds/qTox/local-deps/lib/libsqlcipher.0.dylib (compatibility version 9.0.0, current version 9.6.0)
        libvpx.7.dylib (compatibility version 1.0.0, current version 1.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
        @rpath/libtoxcore.2.dylib (compatibility version 2.0.0, current version 2.13.0)
        libopenal.1.dylib (compatibility version 1.0.0, current version 1.16.0)
        /System/Library/Frameworks/AVFoundation.framework/Versions/A/AVFoundation (compatibility version 1.0.0, current version 2.0.0)
        /System/Library/Frameworks/CoreMedia.framework/Versions/A/CoreMedia (compatibility version 1.0.0, current version 1.0.0)
        /System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1557.3.2)
        /System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1856.105.0)
        /System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
        /usr/local/opt/qt@5/lib/QtGui.framework/Versions/5/QtGui (compatibility version 5.15.0, current version 5.15.2)
        /Users/abilinski/builds/qTox/local-deps/lib/libopus.0.dylib (compatibility version 9.0.0, current version 9.0.0)
        /usr/local/opt/qt@5/lib/QtCore.framework/Versions/5/QtCore (compatibility version 5.15.0, current version 5.15.2)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1200.3.0)
        /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1856.105.0)
        /usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```

Which shows libvpx and libopenal, the two libs causing issues in CI, as not being found in the local-deps dir, despite the rest being found.

anthonybilinski commented:
Fixed the linking issues for libvpx and libopenal, PR should be ready to go. Testing the actual build I found a couple major bugs, but I think those should be addressed separately. For one, toxext negotiates successfully but then thinks the friend's max message size is -1 so fails to send any messages. Secondly, the settings panes all become completely blank once the A/V pane is clicked.

anthonybilinski committed:
chore(CI): Allow cross compile scripts to be called with relative paths

Like was done in e40586803743c9f08a9018caccf8038f754a31e4

anthonybilinski committed:
chore(CI): Add macOS platform detection for build scripts

anthonybilinski committed:
chore(CI): Prefix Windows arch with win to disambiguate with other platforms

anthonybilinski commented:
I'm splitting this up and PRing a series of smaller PRs, so am closing this one.
